ENH: Use `threadpoolctl` in `show_runtime` (a new function)

ganesh-k13 opened the pull request:
### ENH: Use `threadpoolctl`

* Use `threadpoolctl` at ~`np.show_config`~ `np.show_runtime`
* Refactor SIMD extensions display logic

### Sample output
```py
>>> np.show_runtime()
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}},
 {'architecture': 'Zen',
  'filepath': '/usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so',
  'internal_api': 'openblas',
  'num_threads': 12,
  'prefix': 'libopenblas',
  'threading_layer': 'pthreads',
  'user_api': 'blas',
  'version': '0.3.20'}]
```

or

```py
>>> np.show_runtime()
WARNING: `threadpoolctl` not found in system! Install it by `pip install threadpoolctl`. Once installed, try `np.show_runtime` again for more detailed build information
[{'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]
```

resolves: #21340
related: #20939

rgommers commented:
Thanks for working on this @ganesh-k13!

Note that gh-14492 is also very much related. I think we'd like `show_config` to display both build-time and runtime info, and clearly state which is which (right now it's build-time only info, but it's a bit confusing).

charris commented:
@rgommers Out of curiosity, can the build time information be obtained from meson?

ganesh-k13 commented:
> display both build-time and runtime info

Ah ok, thanks for the info @rgommers! Let me see how we can pull up that info today. As @charris suggested, maybe meson can help, but only if it's built using that?

rgommers commented:
> @rgommers Out of curiosity, can the build time information be obtained from meson?

There's always a way:) It has sane configure-time output, which may make sense to store in full:

<details>

```
meson setup build
The Meson build system
Version: 0.61.0
Source dir: /Users/rgommers/code/bldscipy
Build dir: /Users/rgommers/code/bldscipy/build
Build type: native build
Project name: SciPy
Project version: 1.9.0.dev0
C compiler for the host machine: arm64-apple-darwin20.0.0-clang (clang 11.1.0 "clang version 11.1.0")
C linker for the host machine: arm64-apple-darwin20.0.0-clang ld64 530
C++ compiler for the host machine: arm64-apple-darwin20.0.0-clang++ (clang 11.1.0 "clang version 11.1.0")
C++ linker for the host machine: arm64-apple-darwin20.0.0-clang++ ld64 530
Host machine cpu family: aarch64
Host machine cpu: arm64
Compiler for C supports arguments -Wno-unused-but-set-variable: NO
Library m found: YES
['/Users/rgommers/mambaforge/envs/scipy-dev/bin/arm64-apple-darwin20.0.0-gfortran', '-Wno-conversion', '-Wl,-pie', '-Wl,-headerpad_max_install_names', '-Wl,-dead_strip_dylibs', '-Wl,-rpath,/Users/rgommers/mambaforge/envs/scipy-dev/lib', '-L/Users/rgommers/mambaforge/envs/scipy-dev/lib', '/Users/rgommers/code/bldscipy/build/meson-private/sanitycheckf.f90', '-o', '/Users/rgommers/code/bldscipy/build/meson-private/sanitycheckf']
AHH
['/Users/rgommers/code/bldscipy/build/meson-private/sanitycheckf']
 Fortran compilation is working.
0
Fortran compiler for the host machine: /Users/rgommers/mambaforge/envs/scipy-dev/bin/arm64-apple-darwin20.0.0-gfortran (gcc 11.0.1 "GNU Fortran (GCC) 11.0.1 20210403 (experimental)")
Fortran linker for the host machine: /Users/rgommers/mambaforge/envs/scipy-dev/bin/arm64-apple-darwin20.0.0-gfortran ld64 530
Program cython found: YES (/Users/rgommers/mambaforge/envs/scipy-dev/bin/cython)
Program pythran found: YES (/Users/rgommers/mambaforge/envs/scipy-dev/bin/pythran)
Program cp found: YES (/bin/cp)
Program python3 found: YES (/Users/rgommers/mambaforge/envs/scipy-dev/bin/python3.9)
WARNING: Python files installed by Meson might not be found by python interpreter.
 This warning can be avoided by setting "python.platlibdir" option.
WARNING: Python files installed by Meson might not be found by python interpreter.
 This warning can be avoided by setting "python.purelibdir" option.
Found pkg-config: /Users/rgommers/mambaforge/envs/scipy-dev/bin/pkg-config (0.29.2)
Library npymath found: YES
Library npyrandom found: YES
Run-time dependency openblas found: YES 0.3.18
Dependency openblas found: YES 0.3.18 (cached)
Program _build_utils/cythoner.py found: YES (/Users/rgommers/mambaforge/envs/scipy-dev/bin/python3.9 /Users/rgommers/code/bldscipy/scipy/_build_utils/cythoner.py)
Checking for function "open_memstream" : YES
Configuring messagestream_config.h using configuration
Checking for size of "void*" : 8
Run-time dependency threads found: YES
Checking for size of "void*" : 8
Dependency threads found: YES unknown (cached)
Compiler for C supports arguments -Wno-unused-but-set-variable: NO (cached)
Dependency openblas found: YES 0.3.18 (cached)
Compiler for C++ supports arguments -mmacosx-version-min=10.9: YES
Compiler for C supports arguments -Wno-unused-but-set-variable: NO (cached)
Compiler for C++ supports arguments -Wno-format-truncation: NO
```

</details>

It contains `Run-time dependency openblas found: YES 0.3.18`, so yes that covers BLAS/LAPACK at build time, but some of the other things in there are equally useful for debugging user-reported issues.

mattip commented:
I would be fine with fixing `show_config` one step at a time if it is complicated to do all the fixes in one go. Having the `threadpoolctl` runtime information would already help in many situations.

ganesh-k13 commented:
Hey @mattip, I see two enhancements based on various discussions,
1. Show both runtime and build time info.
2. Figure out a fix for Windows.

Which do you think is apt for this PR, I can work on that.

mattip commented on numpy/distutils/misc_util.py:
```diff
@@ -2371,9 +2352,8 @@ def show():
-                directories, include directories, etc.) in the system on which
-                NumPy was built.
+                Print information about various resources in the system
+                on which NumPy was built.
```
The information presented no longer is the build-time information.
```suggestion
                including available intrinsic support and BLAS/LAPACK library
                in use
```

ganesh-k13 replied:
Done via [49ed995](https://github.com/numpy/numpy/pull/21468/commits/49ed9952bffcf55fa4d298e05aee3d18138fe2a3)

mattip commented on numpy/distutils/misc_util.py:
```diff
@@ -2382,84 +2362,58 @@ def show():
+                'threading_layer': 'pthreads',
+                'user_api': 'blas',
+                'version': '0.3.20'}]
+
```
Where is the example of the SIMD information?

ganesh-k13 replied:
Didn't want to clutter the docstring, but yeah makes sense to add it as it'll always be displayed

ganesh-k13 replied:
Done via [49ed995](https://github.com/numpy/numpy/pull/21468/commits/49ed9952bffcf55fa4d298e05aee3d18138fe2a3)

ganesh-k13 commented:
Oh yeah that's a good point, we can make `show_runtime()` and later `show_buildinfo()` (or something like that) and then merge with `show_config` if it's received well. Will do 👍

mattip commented:
I see you updated the PR description to show the new function output. It looks good to me.

This is failing since it adds a new function to the namespace. It should probably hit the mailing list in parallel with fixing the test to allow the addition. When you post, please add the output.

ganesh-k13 commented:
Thanks Matti, fixing the tests now.

I've posted to the mailing list today: https://mail.python.org/archives/list/numpy-discussion@python.org/thread/A3Q2WXBRFK3MKHOSEX3RUVRJRKCFGWKX/

mattip commented:
I don't think this needs to be part of the generated `__config__.py` file. Maybe a better home would be in `numpy/lib/utils.py`. The `distutils` module is due to be deprecated.

ganesh-k13 committed:
ENH: Added `show_runtime`

1. Information is derived with the help of `threadpoolctl` library.
2. In case `threadpoolctl` is not installed, a message is displayed
   with help on how to install it.
3. SIMD related information is derived from `__cpu_features__`,
   `__cpu_baseline__` and `__cpu_dispatch__`

ganesh-k13 committed:
TST: Added show_runtime to test_public_api

ganesh-k13 committed:
DOC: Added `np.show_runtime` (#21468)

ganesh-k13 commented:
Thanks for the review Matti! I have moved it to lib/utils.py and also added a release note.

mattip commented on doc/release/upcoming_changes/21468.new_feature.rst:
```diff
@@ -0,0 +1,60 @@
+A new function ``np.show_runtime`` has been added to display the runtime
+information of the machine in addition to ``np.show_config`` which displays
+the build-related information.
+This function returns a list of dictionaries as follows::
```
I think everything from here down should be removed. `show_runtime` should be added to the documentation here https://github.com/numpy/numpy/blob/58ee8a8bde14a84fa261ae329c3f99d953644f2c/doc/source/reference/routines.other.rst?plain=1#L47. Then the references above could change to `numpy.show_runtime`, and `numpy.show_config` (with single back-ticks), and they would be turned into links.

ganesh-k13 replied:
Wow yeah, that's nice! Added via [706c922](https://github.com/numpy/numpy/pull/21468/commits/706c922ff278a6197f47ba678e8cc6128028d5ab)

mattip commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+    """
+    Print information about various resources in the system
+    including available intrinsic support and BLAS/LAPACK library
+    in use
+    See Also
```
Blank line needed
```suggestion
    in use

    See Also
```

ganesh-k13 replied:
Added in [fe63699](https://github.com/numpy/numpy/pull/21468/commits/fe6369961f4ff136161b0dac2ae2ec620dbce5fb)

mattip commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+    in use
+    See Also
+    --------
+    show_config : Show libraries in the system on which
+                  NumPy was built.
+    Notes
```
Blank line, reformat
```suggestion
    show_config : Show libraries in the system on which NumPy was built.

    Notes
```

ganesh-k13 replied:
Added in [fe63699](https://github.com/numpy/numpy/pull/21468/commits/fe6369961f4ff136161b0dac2ae2ec620dbce5fb)

mattip commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+    2. In case `threadpoolctl` is not installed, a message is
+       displayed with help on how to install it.
+    3. SIMD related information is derived from `__cpu_features__`,
+       `__cpu_baseline__` and `__cpu_dispatch__`
+    Examples
```
Blank line needed
```suggestion
       `__cpu_baseline__` and `__cpu_dispatch__`

    Examples
```

ganesh-k13 replied:
Added in [fe63699](https://github.com/numpy/numpy/pull/21468/commits/fe6369961f4ff136161b0dac2ae2ec620dbce5fb)

mattip commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+    -----
+    1. Information is derived with the help of `threadpoolctl`
+       library.
+    2. In case `threadpoolctl` is not installed, a message is
+       displayed with help on how to install it.
```
Not sure this is really needed

ganesh-k13 replied:
Removed in [fe63699](https://github.com/numpy/numpy/pull/21468/commits/fe6369961f4ff136161b0dac2ae2ec620dbce5fb)

mattip reviewed (commented):
LGTM. Just some documentation tweaks.

ganesh-k13 committed:
DOC: Added `show_runtime` to Utility

mattip commented on doc/release/upcoming_changes/21468.new_feature.rst:
```diff
@@ -0,0 +1,6 @@
+New function `np.show_runtime`
+------------------------------
+
+A new function `np.show_runtime` has been added to display the runtime
+information of the machine in addition to `np.show_config` which displays
+the build-related information.
```
Double back-ticks for `np.*` to format as code. If you want a link, you need to use `numpy.*` with single back-ticks. We tried telling sphinx that `np` is an alias for `numpy`, but couldn't get the links to work.
```suggestion
New function ``np.show_runtime``
--------------------------------

A new function `numpy.show_runtime` has been added to display the runtime
information of the machine in addition to `numpy.show_config` which displays
the build-related information.
```

ganesh-k13 replied:
Thanks for the info! Learning Sphinx on the way :). Fixed via [3d2fe4e](https://github.com/numpy/numpy/pull/21468/commits/3d2fe4ec796be6ae63c482e4c4f74e3c7867ff1d)

mattip commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+    -----
+    1. Information is derived with the help of `threadpoolctl`
+       library.
+    2. SIMD related information is derived from `__cpu_features__`,
+       `__cpu_baseline__` and `__cpu_dispatch__`
```
Double back-ticks to format as code. Single back-ticks for links. I don't think we have links for those attributes.
```suggestion
    2. SIMD related information is derived from ``__cpu_features__``,
       ``__cpu_baseline__`` and ``__cpu_dispatch__``
```

ganesh-k13 replied:
Fixed via [3d2fe4e](https://github.com/numpy/numpy/pull/21468/commits/3d2fe4ec796be6ae63c482e4c4f74e3c7867ff1d)

mattip commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+
+    Notes
+    -----
+    1. Information is derived with the help of `threadpoolctl`
```
If you want this to be a link, you have to provide the url. Otherwise use double back-ticks.

ganesh-k13 replied:
Thanks yeah it renders now: https://output.circle-artifacts.com/output/job/79c7902f-6c77-4b35-8f43-d55ceaee6ff5/artifacts/0/doc/build/html/reference/generated/numpy.show_runtime.html, all though it will give flake8 errors, I'm not sure why adding new line like this did not work: https://github.com/numpy/numpy/blob/17d730ae32f5f60c9c2ca75d202b4e866debd686/numpy/lib/npyio.py#L1846-L1847

ganesh-k13 committed:
DOC: Fixed links for np.show_runtime (#21468)

ganesh-k13 commented on numpy/lib/utils.py:
```diff
@@ -13,9 +13,86 @@
+    Notes
+    -----
+    1. Information is derived with the help of `threadpoolctl
+        <https://pypi.org/project/threadpoolctl/>`_ library.
```
It seems sphinx is not rendering this: https://output.circle-artifacts.com/output/job/03b954b5-25ba-40b5-adfb-51afb9291f75/artifacts/0/doc/build/html/reference/generated/numpy.show_runtime.html

Will take a look

mattip commented:
Thanks @ganesh-k13 this will be helpful going forward
